feat(DatePicker): DatePicker implementation

epr3 opened the pull request:
Closes #379.

Hello,

First of all, I want to say thank you for this awesome library.

I implemented this feature taking inspiration from [https://melt-ui.com/](https://melt-ui.com/), [https://www.bits-ui.com/docs/introduction](https://www.bits-ui.com/docs/introduction) and [https://react-spectrum.adobe.com/react-stately](https://react-spectrum.adobe.com/react-stately/), plus the examples provided on the issue.

The components that I've implemented are:

- DateField
- DateRangeField
- Calendar
- RangeCalendar
- DatePicker
- DateRangePicker

The last two are basically composing the field components, the calendar components and the popover component and are re-exporting them under the DatePicker/DateRangePicker naming.

The components are fully navigable by keyboard and should have all accessibility tags.

I open this as a draft PR because it still needs some stuff added to it.

The things that I did not manage to implement are: adding documentation, styling for the components in Histoire, usage examples, general cleanup and do some refactoring to clean the code and prevent unnecessary re-renders.

I appreciate any feedback and will implement any refactor, bugfix or feature required.

zernonia commented:
WHOAAAAA @epr3 !! I was just started looking into this, and you already have the PR ready!!! I'll review it shortly ya! 💚

sadeghbarati commented:
@epr3 Thanks for your time ❤️ 

Found a source that can guide you to add `month` or `year` selection, the project is also using `@internationalized/date`

- https://www.marigold-ui.io/components/form/calendar#simple-calendar
- https://github.com/marigold-ui/marigold/tree/main/packages/components/src/Calendar

I don't know which one is more convenient

- Popover for selection
- Overlay just like the demo sent above

@zernonia wdyt?

epr3 commented:
@sadeghbarati Thank you for your guidance. 

I'm leaning towards implementing the `year` and `month` selection by providing two new primitives: `CalendarYearSelect` and `CalendarMonthSelect` which on click would either replace the values provided through the `months` v-slot value (and probably rename it to something more intuitive) by keeping an internal state for the view and letting the end-user re-use all the `Calendar` building blocks or providing separate values for dates, months and years and the current view through `v-slot` and letting the user the ability to make the decision on how he wants to display these values.

The only thing that I don't know how I should approach for this feature is what should the values be if `number-of-months` is set to something greater than 1 and `paged-navigation` is set to `false`.
<img width="644" alt="image" src="https://github.com/radix-vue/radix-vue/assets/17836403/82dec702-d3a4-4905-829f-f6de0e4dd70e">

[Shadcn Svelte](https://www.shadcn-svelte.com/docs/components/calendar) seems to have used a couple of selects for the month and year.

What do you think?

sadeghbarati commented:
@epr3 It's nice that you create new Primitives for it ❤️

And also use Vue features like slotProps 💚

> The only thing that I don't know how I should approach for this feature is what should the values be if number-of-months is set to something greater than 1 and paged-navigation is set to false

Can you make a demo of this problem? You can push it in this PR so we could checkout to your branch and see what we can do about it

epr3 commented:
@sadeghbarati Unfortunately, I didn't start implementing this feature yet. 

I want to think this edge case through.

For example, in the screenshot above, what months should be displayed on the Month Select View? (for one month it's: Jan, Feb, etc., for two months and paged navigation it would be: Jan-Feb, Mar-Apr, etc.), but if `paged-navigation` is set to false, the calendar would cycle like so: Jan-Feb, Feb-Mar, Mar-Apr, etc., therefore, the issue with the month display when you would click `CalendarMonthSelect`. 

<img width="454" alt="image" src="https://github.com/radix-vue/radix-vue/assets/17836403/2323cf4c-63c7-47cb-a01c-a6d01bc02012">

One possible solution that I can think of is to display the months like so: Jan, Feb, Mar, etc. and when you click one month it would set the start month on the calendar paged navigation.

Hope the description I provided makes sense.

Thanks!

sondh0127 commented on packages/radix-vue/src/DatePicker/DatePickerCalendar.vue:
```diff
@@ -0,0 +1,36 @@
+
+<template>
+  <CalendarRoot
+    v-slot="{ weekDays, months }"
```
It could be easy for @zernonia to review if we separate to Calendar PR before DatePicker.(my suggession)

epr3 commented:
Hello @zernonia and @sadeghbarati,

I marked this PR as ready for review.

Let me know if there's anything else that needs to be implemented or fixed.

Thank you!

sadeghbarati commented:
@epr3 Again thanks for the time you put into this ❤️ 

---


- I see type error in `radix-vue` build script

- The selected and focused date in `Calendar` and `RangeCalendar` will active the `year` cell background color, that `green` dot is enough I think, If I unfocus (not unselect) the selected date this black bg color will be removed

  ![image](https://github.com/radix-vue/radix-vue/assets/17789047/f01e2427-f8e1-4c82-a054-b03513ca0305)

epr3 commented:
> @epr3 Again thanks for the time you put into this ❤️
> 
> * I see type error in `radix-vue` build script
> * The selected and focused date in `Calendar` and `RangeCalendar` will active the `year` cell background color, that `green` dot is enough I think, If I unfocus (not unselect) the selected date this black bg color will be removed
>   ![image](https://private-user-images.githubusercontent.com/17789047/307604216-f01e2427-f8e1-4c82-a054-b03513ca0305.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg4OTg1NjEsIm5iZiI6MTcwODg5ODI2MSwicGF0aCI6Ii8xNzc4OTA0Ny8zMDc2MDQyMTYtZjAxZTI0MjctZjhlMS00YzgyLWEwNTQtYjAzNTEzY2EwMzA1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI1VDIxNTc0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUxMTA4MjliOTdhYTc2ZDhmNmEzN2U5NGVjMjI4OTBkYTBhYTYwNTc4ZjQ1NjVhNjU2YWNhNzhkNWQwNDYxZWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.dZjkKCfOavHmjUF1GuOgYgHOwEqxw1xLfN1sE-iD-vw)

I've fixed the typing bugs and removed the selected attribute on dates when the calendar is not in the `month` view.

maelp commented:
What's the status of this? And is it going to be included in Shadcn-vue ?

sadeghbarati commented:
> What's the status of this? And is it going to be included in Shadcn-vue ?

Sure thing why not.

---

@epr3 

- Can we go next month by clicking those faded dates in the edge of `Date Picker` and `Date Range Picker`?

  ![image](https://github.com/radix-vue/radix-vue/assets/17789047/c4e51c3d-5ded-4eda-9258-5ecbccba1607)
  
  
- I think it's better to have same style as https://react-spectrum.adobe.com/react-aria/DateRangePicker.html for Range Pickers I mean `data-selection-start="true"` and `data-selection-end="true"`

zernonia commented:
Thanks @epr3 ! It's looking good! I will push some cleanup regarding the folder structure and stuff yaa 😁

zernonia commented:
@epr3  I see `DateField`, `Calendar` were broken if we use `v-model` in the story.

epr3 commented:
@zernonia - could you check the `v-model` with the latest commits?

From my testing, it seems that the ref kind of breaks when you use `=` to attribute one `DateValue` to another. Returning a new one works though.

zernonia commented:
yoo @epr3 ! I was looking through different case by adding more stories, and I noticed 2 bugs.


1. DateField is showing `seconds` if the set `granularity="hour"`
http://localhost:6006/story/packages-radix-vue-src-datefield-story-datefieldgranular-story-vue?variantId=packages-radix-vue-src-datefield-story-datefieldgranular-story-vue-2
![image](https://github.com/radix-vue/radix-vue/assets/59365435/ac507637-1b0c-495d-9e5d-9669e737727b)


2. Calendar setting `preventSelect` still enable user to deselect it .
http://localhost:6006/story/packages-radix-vue-src-calendar-story-calendarchromatic-story-vue?variantId=packages-radix-vue-src-calendar-story-calendarchromatic-story-vue-8
![image](https://github.com/radix-vue/radix-vue/assets/59365435/93f66833-49dc-45aa-88e9-9e8f4b73b31e)

epr3 commented:
Hello @zernonia & @sadeghbarati,

Thank you very much for the help and reviews! 

I have a dilemma regarding this point outlined by @sadeghbarati:
> Layout shift in Calendar by clicking Heading Segment month and year which is not accessible I think using Overlay Component (Popover) is better to show months or years, one Popover which contains month and year logic

By using an Overlay Component(Popover), shouldn't the month and year selection feature, at this point, be handled through **shadcn**, by composing a normal Popover component? 

I'm thinking that making the year and month selection an "opt-in" feature through the `placeholder` `v-model` field would reduce the amount of moving parts inside this component, the component itself would not need to re-export a `Popover` component(-4/5 components to worry about) and I think it would be more in line with the compositional nature of Radix.

I've created an example here (also removing the code pertaining to the month and year select):
https://github.com/epr3/radix-vue/blob/4f45fe931957079e07e483b49a54e01104ad10a9/packages/radix-vue/src/Calendar/story/CalendarPopover.story.vue
 I'm leaning towards the year and month selection to be handled by shadcn, but I can also implement the `Popover` inside the `Calendar` and export all `Popover` components under a `Calendar` branding.

What do you guys think about this?

sadeghbarati commented:
Sorry for causing a dilemma 

I saw `antd DatePicker`

https://ant.design/components/date-picker, and those things come to my mind

Yeah, the opt-in feature is nice 🫡,
Instead of adding another component for it

With this, the user is not limited to using it only inside Popover.

zernonia commented:
Let's go with `opt-in` for Alpha version 😁

epr3 commented:
> @epr3 why remove the `defaultValue` props?

They seem redundant, since the `modelValue` would get populated through the `v-model`.

epr3 commented:
@zernonia - I've added back the `defaultValue` props

I also would need some guidance on this issue if possible: Do you have any idea on how we could properly handle the `DateValue` values inside `ref`? 

A lot of errors pop up due to the fact that Vue seems to drop nested values of `DateValue` (always returning a new `DateValue` doesn't always work), namely the `calendar` field which contains a lot of functions needed by `@internationalized/date` for various comparisons/operations and the package itself is built with ES6 classes which I believe builds `Object`s and populates their prototypes with class methods.

zernonia commented:
@epr3  unfortunately I dont think we can have a solution for now, as it seems like a problem with Vue.

https://www.reddit.com/r/vuejs/comments/xbfa3x/prop_as_a_typescript_class_object/
https://stackoverflow.com/questions/75557531/vue3-unwrap-properties-in-typescript-class

However we can get this PR merged. Absolutely banger job @epr3 !!!! 🤩

maelp commented:
Nice! Looking forward to seeing this in Shadcn! @zernonia will you do a custom styling of the component to replace the old one, or as an entirely new one?

ico85 commented:
I would like to express my gratitude and emphasize the importance of component libraries that are consistently developed with accessibility in mind. I work for a service provider that creates websites for public entities (e.g., websites for municipalities and cities, which are legally required to be accessible!). We are legally obligated to implement guidelines that ensure accessibility. In Germany, these guidelines are known as BITV 2.0 ("Barrierefreie Informationstechnik-Verordnung 2.0"). Your framework and your work are incredibly important. Thank you!

valh1996 commented:
Hello @epr3, how do you pass a DateTime instead of a Date? I need to control the time too.

I see that `radix-vue` exports a `DateRange` type but no `DateTimeRange`.

Could you please give me a hint on how to do this (with an example) ? 
```
import type { DateRange } from 'radix-vue'
```
___
*Congratulations on your work! It's fantastic!*

epr3 commented:
Hello @valh1996,

You would need to use `CalendarDateTime` or `ZonedDateTime`  from the `@internationalized/date` package to be able to manipulate the time.

The `DateRange` type is meant to be used with the date range components: `DateRangeField`/`DateRangePicker`/`RangeCalendar`.

valh1996 commented:
> Hello @valh1996,
> 
> You would need to use `CalendarDateTime` or `ZonedDateTime` from the `@internationalized/date` package to be able to manipulate the time.
> 
> The `DateRange` type is meant to be used with the date range components: `DateRangeField`/`DateRangePicker`/`RangeCalendar`.

Ok, thanks, so if I understand correctly, [the calendar doesn't support the time](https://vcalendar.io/datepicker/time-picker.html#time-picker), so I have to manage it myself?

epr3 commented:
Yes, you would need to build the time picker yourself.

sadeghbarati commented:
Will send a Stackblitz demo with description, but with a single calendar date
